crates/valence_event/src/state_event.rs

```diff
@@ -0,0 +1,591 @@
+use std::cmp::Ordering;
+use std::collections::HashMap;
+use std::fmt;
+use std::iter::Chain;
+use std::marker::PhantomData;
+use std::ops::{Deref, DerefMut};
+use std::slice::Iter;
+
+use bevy_ecs::prelude::Event;
+use bevy_ecs::system::{Local, Res, ResMut, Resource, SystemParam};
+
+pub struct EventId<E: Event> {
+    /// Uniquely identifies the event associated with this ID.
+    // This value corresponds to the order in which each event was added to the world.
+    pub id: usize,
+    _marker: PhantomData<E>,
+}
+
+impl<E: Event> Copy for EventId<E> {}
+impl<E: Event> Clone for EventId<E> {
+    fn clone(&self) -> Self {
+        *self
+    }
+}
+
+impl<E: Event> fmt::Display for EventId<E> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        <Self as fmt::Debug>::fmt(self, f)
+    }
+}
+
+impl<E: Event> fmt::Debug for EventId<E> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(
+            f,
+            "event<{}>#{}",
+            std::any::type_name::<E>().split("::").last().unwrap(),
+            self.id,
+        )
+    }
+}
+
+impl<E: Event> Eq for EventId<E> {}
+impl<E: Event> PartialEq for EventId<E> {
+    fn eq(&self, other: &Self) -> bool {
+        self.id == other.id
+    }
+}
+
+impl<E: Event> PartialOrd for EventId<E> {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl<E: Event> Ord for EventId<E> {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.id.cmp(&other.id)
+    }
+}
+
+/// A unique identifier for a state usable by multiple events types.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub struct StateId(usize);
+
+/// Defines a struct that can be used as a state for an event.
+pub trait State: Send + Sync + 'static {
+    fn get(&self) -> bool;
+
+    fn set(&mut self, value: bool);
+}
+
+/// Resource that holds all states for all events.
+#[derive(Resource)]
+pub struct States {
```

Comment:
Bevy already has a type called `States`, a trait used for general game states.
I would recommend renaming this to `EventStates`, which is more obvious anyways.